### ISSUE TITLE
tmux mouse scrolling might not be working right for vim?

### ISSUE THREAD

**Issue by NHDaly** [NHDaly/tmux-scroll-copy-mode#21]
Follow-up conversation from #19:

From @zeorin's [comment](https://github.com/NHDaly/tmux-better-mouse-mode/issues/19#issuecomment-247240871):

> When I start Vim using `vim -u NONE` (don't load plugins or read any initialization files), there's no scrolling at all. When I then run `:set mouse=a`, it scrolls by moving the cursor and only the file as a whole when the top or bottom is reached. My Vim version is 7.4, patches 1-1689 on my Ubuntu machine.

Hmm, that is definitely not how it should be working. We can try to figure out what's wrong.

> Perhaps you have a setting in your `.vimrc` (other than `:set mouse=a`) which effects whether scrolling is accepted? Or a terminal override in your `.tmux.conf`?

Hmm. I do not think this is being triggered by a setting in [my `.vimrc`](https://github.com/NHDaly/nhdaly_dotfiles/blob/master/vimrc.nhdaly): 

For me (see below for my setup), starting in `vim -u NONE` exhibits the following behavior, which matches my expectations:
- At first, when `:set mouse?` says mouse is disabled, scrolling the mouse wheel moves the cursor, thanks to `set -g @emulate-scroll-for-no-mouse-alternate-buffer "on"` in my `tmux.conf`.
- Then, when I `:set mouse=a`, scrolling the mouse wheel moves the background and keeps the cursor in place, because now vim is interpreting my scroll events.

I'm seeing that behavior on my two machines:
1. macbook pro running OSX 10.11.6, iterm2, zsh, vim 7.3
2. Ubuntu (ssh'd from iterm2 on my mac), zsh, vim 7.4 with patches: 1-1816)

---

I also don't see anything in [my `.tmux.conf`](https://github.com/NHDaly/nhdaly_dotfiles/blob/master/tmux.conf.nhdaly) that looks suspicous.

It could be something in [my `.zshrc`](https://github.com/NHDaly/nhdaly_dotfiles/blob/master/zshrc.nhdaly), but that looks innocent too...

---

But that doesn't leave us with much. Could it be that your `Dash` isn't forwarding the mouse events in the same way as my zsh is?

How does the mouse-mode behave if you open `vim` in a regular terminal pane outside of `tmux`?

**Comment by zeorin** [comment]
`vim` in a regular terminal pane outside `tmux` behaves properly, i.e. it responds to the mouse scroll events.

I figured it out: it has to do with the `ttymouse` option in Vim (NeoVim doesn't have this option). Normally, Vim automatically sets the `ttymouse` option depending on the value of `$TERM`. The default value of `$TERM` inside a tmux pane is `screen-256color` (at least, it is for me). This is a value that Vim recognizes.

However, I've changed my `$TERM` inside a tmux pane to `tmux-256color`, because it was not handling italics correctly otherwise (for which I had to add a TERMINFO entry, as per the tmux FAQ). This is not a value that Vim recognizes, and thus it leaves `ttymouse` unset.

When `ttymouse` is unset, the value of `#{mouse_any_flag}` in tmux is `0`, and thus your plugin sends arrow keys instead of mouse scroll events.

Doing a simple check in my `.vimrc` to see whether I'm a) running in Vim and not NeoVim, and b) `$TERM` is set to `tmux-256color`; and then setting `ttymouse` to `xterm` (not `urxvt`, even though that's my terminal—tmux largely emulates xterm), fixes the problem and Vim now scrolls like NeoVim.

So this issue was not caused at all by your plugin! Thanks very much for your questions, though: they led me to investigate the right avenues.
